metals/src/main/scala/scala/meta/internal/metals/BSPErrorHandler.scala

```diff
@@ -0,0 +1,174 @@
+package scala.meta.internal.metals
+
+import java.nio.file.Files
+import java.util.concurrent.atomic.AtomicReference
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+import scala.util.control.NonFatal
+
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.metals.clients.language.MetalsLanguageClient
+import scala.meta.io.AbsolutePath
+
+import org.eclipse.{lsp4j => l}
+
+class BSPErrorHandler(
+    languageClient: MetalsLanguageClient,
+    workspaceFolder: AbsolutePath,
+)(implicit context: ExecutionContext) {
+
+  private def logsPath = workspaceFolder.resolve(Directories.log)
+  private val currentError =
+    new AtomicReference[CurrentAndPreviousError](new CurrentAndPreviousError())
+
+  def onError(message: String): Unit = {
+    currentError.getAndUpdate { collectedError =>
+      if (collectedError.isEmpty) {
+        showErrorWithDelay()
```

Comment:
We show error message inside the `getAndUpdate` and the doc says `The function should be side-effect-free, since it may be re-applied when attempted updates fail due to contention among threads.` Should we care about the possibility of showing multiple error messages, or doing it inside a `Future` somehow helps?

Reply:
Yeah, I think you're right I have to take a different approach to make sure this works as it should.

Reply:
We are also doing `getAndUpdate` inside another `getAndUpdate` on the same resource, which I think might be a problem